### ISSUE TITLE
Unable to detect BeautifulSoup html5lib parser

### ISSUE THREAD

**Issue by sybo-matthew** [tweag/FawltyDeps#246]
**Describe the bug**
FawltyDeps will recommend the removal of the HTML5 parser used by BeautifulSoup

**To Reproduce**
- install beautifulsoup
- set beautifulsoup to use the html5lib parser
- install the html5lib parser with pip install html5lib 
- run fawltydeps and it will recommend the removal of the library as it is not called by import
- attempt to run the script by following fawltydeps recommendations and it will not work as beautifulsoup cannot find the parser it needs

**Expected behavior**
- either to detect that the parser has been called and include it in the list of requirements or to add it to a package blacklist where it'll be ignored for removal considerations. 

**Environment**
 -  macOS Venture 13.2.1  Python 3.11.2 FawltyDeps v0.7.0

If needed I can make a simple test project to display the behavior as well.

**Comment by Nour-Mws** [comment]
Thanks for reporting this @sybo-matthew!

We have an item on our short to medium term roadmap #195 to enhance imports parsing in non straightforward cases. Though this particular case has not been brought to our attention before.

In the meantime, you can run fawltydeps with the `--ignore-unused html5lib` option. 

## Steps to reproduce
1. Python project with `requirements.txt` and `script.py`:
```
# requirements.txt
beautifulsoup4
html5lib
```

```python
# script.py
from bs4 import BeautifulSoup
BeautifulSoup("<a></p>", "html5lib")

```

2. A Python environment `soup_env` where the above requirements are installed.

3. Run fawltydeps 0.7.0 as follows:
```shell
fawltydeps bsoup_project --pyenv soup_env
```

**Result**
```
These dependencies appear to be unused (i.e. not imported):
- 'html5lib'

For a more verbose report re-run with the `--detailed` option.
```

**Comment by jherland** [comment]
I took a look at beautifulsoup itself, and it seems that since v4.4.0 (July 2015), it has provided an `html5lib` "extra" that pulls in `html5lib` as a dependency of `beautifulsoup4`. In other words: if you install `beautifulsoup4[html5lib]` _instead_ of just `beautifulsoup`, you will get `html5lib` automatically without having to install it separately. Since you can now remove your direct dependency on `html5lib`, FawltyDeps will no longer report it as unused.

This seems not to be advertised in beautifulsoup's own documentation, which IMHO is not ideal.

As @Nour-Mws already stated, FawltyDeps still needs to improve in this area, and its report of unused dependencies should definitely _not_ be followed blindly, but should rather be evaluated by someone that knows the project.

**Comment by Nour-Mws** [comment]
@jherland good catch! It seems beutifulsoup's [documentation](https://tedboy.github.io/bs4_doc/2_installation.html#parser-installation) suggests installing parsers into the development environment, which as you said is not ideal.

In this case, the problem was not that `html5lib` was not imported in the Python code but that beautifulsoup4 was installed without it. I suspect running the current version of FawltyDeps on beautifulsoup4 would report html5lib as undeclared.

**Comment by jherland** [comment]
Actually, since current `beautifulsoup4` declares `html5lib` as an "extra" dependency, I think running FawltyDeps on `beautifulsoup4` would match the import against that, and everything would be considered "fine".

**Comment by Nour-Mws** [comment]
@jherland can you link the beautifulsoup4 code you're looking at?

**Comment by jherland** [comment]
Line 32 here: https://git.launchpad.net/beautifulsoup/tree/setup.py#n32

**Comment by jherland** [comment]
@Nour-Mws: FYI, I started working on adding this example to FD under `tests/sample_projects`.

**Comment by Nour-Mws** [comment]
Closing this issue. The immediate problem seems to be unblocked. And the roadmap links back to this issue.
@sybo-matthew feel free to update if you need more information. Thanks again for reporting.
